Rewrite of the installation script and accompanying documentation.

dolfs opened the pull request:
I have redone the prep.sh script functionality making lots of improvements with respect to installing prerequisites, building libavutil et all. as necessary, checking for minimum needed versions of various packages and causing updates to be installed as necessary.
Also updated the documentation to reflect the new usage (there are now several scripts, kept in a separate directory).

scottlamb commented on yarn.lock:
```diff
@@ -1,3510 +0,0 @@
-# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
```
My understanding is it's a best practice to have this file checked in (and the analogous Cargo.lock for the server-side stuff) for reproducible build results.

https://yarnpkg.com/blog/2016/11/24/lockfiles-for-all/

scottlamb commented on scripts/webpack.sh:
```diff
@@ -0,0 +1,8 @@
+#!/bin/sh
+#
+webpack
+RESULT=$?
```
maybe do "#!/bin/sh -e" instead? That would accomplish the same thing, and also complain if the ln command fails.

scottlamb commented on guide/easy_install.md:
```diff
@@ -0,0 +1,156 @@
+There are a few things to prepare if you want a truly turnkey install, but
+they are both optional.
+
+### Cameras.sql preparation
```
I think now that there's "moonfire-nvr config", it's much easier to go through those menus than prepare this file. I tried to remove all references to it when adding that command, but I see now I missed one in .gitignore and one in guide/install.md.

scottlamb commented on guide/easy_install.md:
```diff
@@ -0,0 +1,156 @@
+to issue the following commands:
+
+    $ cd /dev/disk/by-uuid
+    $ ls -l
```
nit: maybe just one command, ls -l /dev/disk/by-uuid?

scottlamb commented on guide/easy_install.md:
```diff
@@ -0,0 +1,156 @@
+will be created and the script will stop. This file is where you will set
+or change variables that describe the moonfire installation you want. The
+initial execution will put default values in this value, but only for the
+most commonly changed variabled. For a full list of variables, see below.
```
s/variabled/variables/

scottlamb commented on guide/easy_install.md:
```diff
@@ -0,0 +1,156 @@
+
+There are two options you may pass to this script. The first is "-B" which
+means "build only". In other words, this will stop the automatic invocation
+of the instal script. The other option available is "-t" and it causes the
```
s/instal/install/

scottlamb replied:
s/and it causes/and causes/

scottlamb commented on guide/easy_install.md:
```diff
@@ -0,0 +1,156 @@
+means "build only". In other words, this will stop the automatic invocation
+of the instal script. The other option available is "-t" and it causes the
+script to ignore the results of any tests. In other words, even if tests
+fail, the build phase will be considered succesful. This can occasionally
```
s/succesful/successful/

scottlamb commented on guide/easy_install.md:
```diff
@@ -0,0 +1,156 @@
+
+This script will copy various files resulting from the build to the correct
+locations. It will also create a "service configuration" for systemctl that
+can be used to control moonfire. This service confuration can be prevented
```
s/confuration/configuration/

scottlamb commented on guide/easy_install.md:
```diff
@@ -0,0 +1,156 @@
+
+Although not all listed in the default prep.config file, these are the
+available configuration variable and their defaults. In the most frequent
+scenarios you will probably only change the port to 80 (assuming you
```
Does changing this to 80 work? Ports < 1024 are privileged, and I don't see anything here that would allow the moonfire-nvr user to bind to that port.

Eventually I hope to support passing the already-bound socket from systemd, which I gather is what the cool kids do. For now, I just use port 8080 in my own setup.

scottlamb commented on guide/install-manual.md:
```diff
@@ -0,0 +1,202 @@
+  servers and write `.mp4` files).
+
+  Note ffmpeg library versions older than 55.1.101, along with all versions of
+  the competing project [libav](http://libav.org), don't not support socket
```
I think this is my typo, but as long as you're touching this, could you s/don't not support/don't support/?

scottlamb commented on guide/install.md:
```diff
@@ -108,7 +41,8 @@ state:
 Both kinds of state are intended to be accessed only by Moonfire NVR itself.
 However, the interface for adding new cameras is not yet written, so you will
```
Oops, I forgot to update this. You can add cameras through "moonfire-nvr config" now.

scottlamb commented on guide/install.md:
```diff
@@ -86,6 +85,10 @@ There are several reasons this is needed:
      downloading it), it stays around until the file is closed. Moonfire NVR
      currently doesn't account for this.
 
+The sample files directory can be set or changed using:
```
This is misleading. Currently, on trunk, you need to pass in the same `--sample-file-dir` arg to every invocation of both config and run. The only time you should change the value is if you've manually moved the entire directory contents between invocations.

If you ever specify the wrong path, it can get confused—some files written in one dir, some in the other, some files that it thinks are deleted but really aren't. And the best process for recovering is basically then copying/moving everything from one dir to the other (assuming you have disk space) and deleting all the files "moonfire-nvr config" says shouldn't exist. It's awkward and undocumented.

I think it's best to just not mess with the default storage location unless you have a good reason. You can set up a symlink to another mount point.

Some of the `new-schema` branch changes are intended to make this safer. The path is stored in the database (paths, actually; you can have multiple dirs to support multiple hard drives/volumes), and some metadata is stored in a file in this directory, and rather than getting confused if you screw up, it will complain if the metadata is absent or doesn't match the database. I still need to work out a UI for fixing it if you've moved the directory or made some other mistake, but the safety checks stop it from making the mess bigger.

scottlamb commented on guide/install.md:
```diff
@@ -32,27 +32,26 @@ state:
     $ sudo adduser --system moonfire-nvr --home /var/lib/moonfire-nvr
     $ sudo mkdir /var/lib/moonfire-nvr
-    $ sudo -u moonfire-nvr -H mkdir db sample
+    $ sudo -u moonfire-nvr -H mkdir db samples
```
sample was correct; this matches the flag default.

scottlamb commented on guide/install.md:
```diff
@@ -86,6 +85,10 @@ There are several reasons this is needed:
 
+The sample files directory can be set or changed using:
+
+    $ moonfire-nvr config --samples-dir=/path/to/samples/dir
```
It's --sample-file-dir.

dolfs commented:
* Put prep.config back into its “proper” location in .gitignore
* Removed yarn.lock from .gitignore; restored prior yarn.lock
* web pack.sh modified as requested
* Made various text corrections
* Updated guide/install-manual to remove references to cameras.sql in favor of "moonfire-nvr" config command
* Updated guide/install-manual to remove references to prep.sh as this is now contained in the easy-install.md file
* Updated/improved documentation further
* Modified setup script to have default values in only one location
* Modified install script to handle privileged port use (system configuration)

Note one final addition: If the user uses a privileged port number the install script prepares the system setup to handle that. The server initialization needs a small modification to also handle it because an "open" socket file descriptor will be inherited in this case. Some example documentation: [https://www.darkcoding.net/software/systemd-socket-activation-in-go/](https://www.darkcoding.net/software/systemd-socket-activation-in-go/)

Until this code modification has been made (scott?) privileged port numbers will still not work, as they have not until now.

scottlamb commented on guide/install.md:
```diff
@@ -32,27 +32,26 @@ state:
    * a SQLite database, typically <1 GiB. It should be stored on flash if
      available.
-   * the "sample file directory", which holds the actual samples/frames of
+   * the "sample files directory", which holds the actual samples/frames of
```
"sample file directory" is the plurality used in the flag/code.

dolfs commented:
Done.

dolfs commented:
You commented on the sample-file-dir option and how it should not be messed with and to use symbolic links instead. I have pretty much removed the remark about it, but the script will still set it to whatever directory was setup in prep.config.
I personally do not like symbolic links for this kind of purpose.

Secondarily I think it would be a good change to basically allow the default directory to be compiled into moonfire, and ditto for db. Doing so would have to change the defaults in the usage command, but would pretty much remove the need to ever issue the sample and db dir options foray given user. Thus also helping to reduce the issue.

scottlamb commented on scripts/setup-ubuntu.sh:
```diff
@@ -0,0 +1,263 @@
+# Now make sure we have dev environment and tools for the UI portion
+#
+echo "Adding yarn components we need..."; echo
+grep -qv \"webpack\" package.json && yarn add --dev webpack@${WEBPACK_MIN_VERSION}
```
This manipulates the checked-in package.json and yarn.lock. I'd rather we do that directly than have people run scripts that do so. Let's remove this and the matching lines in script-functions.sh.

scottlamb commented on scripts/install.sh:
```diff
@@ -0,0 +1,148 @@
+fi
+
+
+# Prepare service files for socket when using priviliged port
```
Could we skip this part for now, and the suggestion in the guide to use port 80? Having it here suggests that it will work. From your link, it sounds not terribly hard to implement but it's not quite at the top of my list. (If you'd like to do it, I'd be happy to review/advise though.)

If anything we should just have if [ $NVR_PORT -lt 1024 ]; then echo "error: unsupported..." 2>&1; ...; fi

scottlamb commented on guide/easy-install.md:
```diff
@@ -0,0 +1,146 @@
+main file systems have a limited lifetime and are way too small to hold
+any significant amount of video.
+If a dedicated hard drive is available, set up the mount point (in this 
+example we'll use /media/nvr/samples):
```
could everything throughout use the singular term "sample file dir" as opposed to "samples dir" or "samples media dir"? (here, install-manual.sh, the script-functions.sh, build.sh.) just seems less confusing to be consistent; less likely people will misremember the commandline flag name and such then.

scottlamb commented on ui-src/index.js:
```diff
@@ -53,15 +53,51 @@ function req(url) {
-  const m = moment.tz(ts90k / 90, zone);
-  const frac = ts90k % 90000;
-  return m.format('YYYY-MM-DDTHH:mm:ss:' + String(100000 + frac).substr(1) + 'Z');
+/**
```
Looks like the changes from your other PR got added here somehow.

scottlamb commented on guide/easy-install.md:
```diff
@@ -129,8 +129,7 @@ start of this configuration.
 available configuration variable and their defaults. In the most frequent
-scenarios you will probably only change the port to 80 (assuming you
-have a dedicated server for moonfire), and SAMPLES_MEDIA_DIR to point
+scenarios you will probably only change SAMPLES_MEDIA_DIR to point
```
These variable names are still plural. SAMPLES_MEDIA_DIR seems like a whole different term, too. SAMPLE_FILE_DIR?

dolfs commented:
Hopefully we're done now?

scottlamb commented:
yep. :) Thanks!
